libsql-sqlite3/src/vectordiskann.c

```diff
@@ -345,4 +403,1048 @@ void nodeBinDebug(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot) {
+  if( vectorIdxKeyNamesRender(pKey->nKeyColumns, "index_key", columnSqlNames, sizeof(columnSqlNames)) != 0 ){
+    rc = SQLITE_ERROR;
+    goto out;
+  }
+  zSql = sqlite3MPrintf(pIndex->db, "SELECT %s FROM %s WHERE rowid = ?", columnSqlNames, pIndex->zShadow);
+  if( zSql == NULL ){
+    rc = SQLITE_NOMEM;
+    goto out;
+  }
+  rc = sqlite3_prepare_v2(pIndex->db, zSql, -1, &pStmt, 0);
+  if( rc != SQLITE_OK ){
+    goto out;
+  }
+  rc = sqlite3_bind_int64(pStmt, 1, nRowid);
+  if( rc != SQLITE_OK ){
+    goto out;
+  }
+  rc = sqlite3_step(pStmt);
+  if( rc != SQLITE_ROW ){
+    goto out;
+  }
+  for(i = 0; i < pRows->nCols; i++){
+    rc = vectorOutRowsPut(pRows, iRow, i, NULL, sqlite3_column_value(pStmt, i));
+    if( rc != SQLITE_OK ){
+      goto out;
+    }
+  }
+
+  // check that we has only single row matching the criteria (otherwise - this is a bug)
+  assert( sqlite3_step(pStmt) == SQLITE_DONE );
+  rc = SQLITE_OK;
+out:
+  if( pStmt != NULL ){
+    sqlite3_finalize(pStmt);
+  }
+  if( zSql != NULL ){
+    sqlite3DbFree(pIndex->db, zSql);
+  }
+  return rc;
+}
+
+/*
+ * Insert new empty row to the shadow table and set new rowid to the pRowid (data will be zeroe-filled blob of size pIndex->nBlockSize)
+*/
+static int diskAnnInsertShadowRow(const DiskAnnIndex *pIndex, const VectorInRow *pVectorInRow, u64 *pRowid){
+  int rc, i;
+  sqlite3_stmt *pStmt = NULL;
+  char *zSql = NULL;
+
+  char columnSqlPlaceholders[DISKANN_SQL_RENDER_LIMIT]; // just placeholders (e.g. ?,?,?, ...)
+  if( vectorInRowPlaceholderRender(pVectorInRow, columnSqlPlaceholders, sizeof(columnSqlPlaceholders)) != 0 ){
+    rc = SQLITE_ERROR;
+    goto out;
+  }
+  zSql = sqlite3MPrintf(pIndex->db, "INSERT INTO %s VALUES (%s, ?) RETURNING rowid", pIndex->zShadow, columnSqlPlaceholders);
+  if( zSql == NULL ){
+    rc = SQLITE_NOMEM_BKPT;
+    goto out;
+  }
+  rc = sqlite3_prepare_v2(pIndex->db, zSql, -1, &pStmt, 0);
+  if( rc != SQLITE_OK ){
+    goto out;
+  }
+  for(i = 0; i < pVectorInRow->nKeys; i++){
+    rc = sqlite3_bind_value(pStmt, i + 1, vectorInRowKey(pVectorInRow, i));
+    if( rc != SQLITE_OK ){
+      goto out;
+    }
+  }
+  rc = sqlite3_bind_zeroblob(pStmt, pVectorInRow->nKeys + 1, pIndex->nBlockSize);
+  if( rc != SQLITE_OK ){
+    goto out;
+  }
+  rc = sqlite3_step(pStmt);
+  if( rc != SQLITE_ROW ){
+    rc = SQLITE_ERROR;
+    goto out;
+  }
+  *pRowid = sqlite3_column_int64(pStmt, 0);
+
+  // check that we has only single row matching the criteria (otherwise - this is a bug)
+  assert( sqlite3_step(pStmt) == SQLITE_DONE );
+  rc = SQLITE_OK;
+out:
+  if( pStmt != NULL ){
+    sqlite3_finalize(pStmt);
+  }
+  if( zSql != NULL ){
+    sqlite3DbFree(pIndex->db, zSql);
+  }
+  return rc;
+}
+
+/*
+ * Delete row from the shadow table
+*/
+static int diskAnnDeleteShadowRow(const DiskAnnIndex *pIndex, i64 nRowid){
+  int rc;
+  sqlite3_stmt *pStmt = NULL;
+  char *zSql = sqlite3MPrintf(pIndex->db, "DELETE FROM %s_shadow WHERE rowid = ?", pIndex->zName);
```

Comment:
Would it make sense to use pIndex->zShadow and %s here instead of pIndex->zName and %s_shadow?